primitives/src/crh/mod.rs

```diff
@@ -23,20 +23,211 @@ pub trait FixedLengthCRH {
+    use crate::{FieldBasedHash, FieldBasedHashParameters};
+
+    struct DummyMNT4BatchPoseidonHash;
+
+    impl BatchFieldBasedHash for DummyMNT4BatchPoseidonHash {
+        type Data = MNT4753Fr;
+        type BaseHash = MNT4PoseidonHash;
+    }
+
+    pub(crate) fn constant_length_field_based_hash_test<H: FieldBasedHash>(
+        digest: &mut H,
+        inputs: Vec<H::Data>,
+    ) {
+        let inputs_len = inputs.len();
+
+        let final_elem = inputs[inputs_len - 1].clone();
+
+        digest.reset(None);
+        inputs.into_iter().take(inputs_len - 1).for_each(|fe| { digest.update(fe); });
+
+        // Test call to finalize() with too few inputs with respect to the declared size
+        // results in an error.
+        assert!(digest.finalize().is_err(), "Success call to finalize despite smaller number of inputs");
+
+        //Test finalize() being idempotent
+        digest.update(final_elem);
+        let output = digest.finalize().unwrap();
+        assert_eq!(output, digest.finalize().unwrap(), "Two subsequent calls to finalize gave different results");
+
+        // Test call to finalize() with too much inputs with respect to the declared size
+        // results in an error.
+        digest.update(final_elem);
+        assert!(digest.finalize().is_err(), "Success call to finalize despite bigger number of inputs");
+    }
+
+    pub(crate) fn variable_length_field_based_hash_test<H: FieldBasedHash>(
+        digest: &mut H,
+        inputs: Vec<H::Data>,
+        mod_rate: bool,
+    ) {
+        let rate = <H::Parameters as FieldBasedHashParameters>::R;
+
+        let pad_inputs = |mut inputs: Vec<H::Data>| -> Vec<H::Data> {
+            inputs.push(H::Data::one());
+            inputs.append(&mut vec![H::Data::zero(); rate - (inputs.len() % rate)]);
+            inputs
+        };
+
+        if mod_rate {
+            constant_length_field_based_hash_test(digest, inputs);
+        } else {
+
+            // Check padding is added correctly and that the hash is collision free when input
+            // is not modulus rate
+            let output = digest.finalize().unwrap();
+            let padded_inputs = pad_inputs(inputs.clone());
+            digest.reset(None);
+            padded_inputs.iter().for_each(|fe| { digest.update(fe.clone()); });
+            assert_ne!(output, digest.finalize().unwrap(), "Incorrect padding: collision detected");
+
+            // Check padding is added correctly and that the hash is collision free also when input
+            // happens to be modulus rate
+            let output = digest.finalize().unwrap();
+            let padded_inputs = pad_inputs(padded_inputs);
+            digest.reset(None);
+            padded_inputs.into_iter().for_each(|fe| { digest.update(fe); });
+            assert_ne!(output, digest.finalize().unwrap(), "Incorrect padding: collision detected");
+        }
+    }
+
+    //#[ignore]
+    #[test]
+    fn test_default_batch_hash_implementation() {
```

Comment:
Actually it rather tests the correctness of the batch_evaluate() in the MNT4BatchPoseidonHash?
The default implementation does just perform the hash evaluations sequentially.

Reply:
It's a simple consistency test of the default implementation of the batch hash